engine/util/src/main/scala/pl/touk/nussknacker/engine/util/cache/DefaultCache.scala

```diff
@@ -0,0 +1,79 @@
+package pl.touk.nussknacker.engine.util.cache
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import com.github.benmanes.caffeine.cache.stats.ConcurrentStatsCounter
+import scalacache.Entry
+
+import scala.concurrent.duration.Duration
+
+/**
+  * @param maximumSize the maximum elements number can contain cache
+  * @param expireAfterAccess the expiration time from last action (read / write)
+  * @tparam T
+  */
+class DefaultCache[T](maximumSize: Long, expireAfterAccess: Option[Duration], registerStats: Boolean) extends Cache[T] {
+
+  import scalacache.caffeine._
+  import scalacache.modes.sync._
+
+  import scala.concurrent.duration._
+
+  private val stats = new ConcurrentStatsCounter
+
+  private lazy val caffeineClientCache = {
+    val builder = Caffeine
+      .newBuilder()
+      .maximumSize(maximumSize)
+      .weakValues()
+
+    expireAfterAccess.foreach(expire => builder.expireAfterAccess(java.time.Duration.ofMillis(expire.toMillis)))
+
+    builder.build[String, Entry[T]]
+  }
+
+  implicit private val cache: CaffeineCache[T] =
+    CaffeineCache(caffeineClientCache)
+
+  override def getOrCreate(key: String, ttl: Option[Duration])(value: => T): T = {
+    if (ttl.exists(time => expireAfterAccess.exists(_.lt(time)))) {
+      throw new IllegalArgumentException(s"TTL (${ttl.get}) should be lower than ExpireAfterAccess (${expireAfterAccess.get}).")
+    }
+
+    cache
+      .doGet(key)
```

Comment:
This approach won't work correctly because of lack of "cirtical section" wrapping `get` and `put`. Consider situation:
1. Thread A doing doGet - return None
2. Thread B doing doGet - return None again
3. Thread A doing doPut
4. Thread B doing doPut
Shouldn't you use: https://cb372.github.io/scalacache/docs/memoization.html instead?

Reply:
Look now.